unittest/Dialect/TritonGPU/DialectTest.cpp

```diff
@@ -518,5 +518,67 @@ TEST_F(InferLayoutTest, FuzzReshape) {
          100.0 * numSuccess / numTests);
 }
 
+std::string toString(Attribute attr) {
```

Comment:
Not sure we want to do C++ unit test for this. The common way to test IR verifications are writing IR tests.  You can write IR tests in `test/TritonGPU/invalid.mlir`.

Reply:
I personally think this test is actually more important than new lit test, because it checks a lot of corner case sin compact form. We already have a lot of lit tests that uses dot encoding already, plus now we use MLIR formatting, which, I believe, is tested in llvm project.

Reply:
I understand your point of checking corner cases and totally agree with that. lit test can do that. :) Per my understanding all the existing lit tests are using the correct/passing paramaters? Don't think we have existing lit tests checking wrong/invalid cases? it's a fairly common practice in mlir to write lit tests for failing validation cases; if you search upstream mlir repo you'll find lots of such examples, like https://github.com/llvm/llvm-project/blob/main/mlir/test/Dialect/Vector/invalid.mlir and various other invalid.mlir files. 

> plus now we use MLIR formatting, which, I believe, is tested in llvm project.

there are autogenerated code to make sure basic type and structure is correct when parsing. and additional constraints are met if you encode in the tablegen. though for validation rules added in C++ code, there will no autogenerated logic for that. need to check by ourselves. so rule of thumb is to add lit test for additional c++ validation rules we add. hopefully this clarifies!